File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/LargestPalindromicSubsequence.ts
function largestPalindromicSubseq(str: string): number {
	const n = str.length;
	const dp: number[][] = new Array(n);

	for (let i = 0; i < n; i++) {
		dp[i] = new Array(n).fill(0);
		dp[i][i] = 1; // each char is itself palindromic with length 1
	}

	for (let l = 1; l < n; l++) {
		for (let i = 0, j = l; j < n; i++, j++) {
			if (str.charAt(i) === str.charAt(j))
				dp[i][j] = dp[i + 1][j - 1] + 2;
			else
				dp[i][j] = Math.max(dp[i + 1][j], dp[i][j - 1]);
		}
	}
	return dp[0][n - 1];
}

const str = "ABCAUCBCxxCBA";
console.log("Largest Palindromic Subseq: " + largestPalindromicSubseq(str));

/*
Largest Palindromic Subseq: 9 
*/

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/LargestPalindromicSubsequence.swift
import Foundation
// Palindromic Subsequence

func largestPalindromicSubseq(_ str : String) -> Int {
	let n : Int = str.count
	var dp : [[Int]] = Array(repeating: Array(repeating: 0, count: n), count: n)
	var i : Int = 0
	while (i < n) {
		dp[i][i] = 1 // each char is itself palindromic with length 1
		i += 1
	}
	
	var l : Int = 1
	while (l < n) {
		var i : Int = 0, j : Int = l
		while (j < n) {
			if (Array(str)[i] == Array(str)[j]) {
				dp[i][j] = dp[i + 1][j - 1] + 2
			} else {
				dp[i][j] = max(dp[i + 1][j],dp[i][j - 1])
			}
			i += 1
			j += 1
		}
		l += 1
	}
	return dp[0][n - 1]
}

// Testing code.
let str : String = "ABCAUCBCxxCBA"
print("Largest Palindromic Subseq:", largestPalindromicSubseq(str))

// Largest Palindromic Subseq: 9